src/GPUActivityAgent.cpp

```diff
@@ -0,0 +1,485 @@
+                                           agent_domain,
+                                           domain_idx), NAN});
+            m_gpu_utilization.push_back({m_platform_io.push_signal("GPU_UTILIZATION",
+                                         agent_domain,
+                                         domain_idx), NAN});
+        }
+
+        for (int domain_idx = 0; domain_idx < m_agent_domain_count; ++domain_idx) {
+            m_gpu_freq_min_control.push_back(control{m_platform_io.push_control("GPU_CORE_FREQUENCY_MIN_CONTROL",
+                                                     agent_domain,
+                                                     domain_idx), NAN});
+            m_gpu_freq_max_control.push_back(control{m_platform_io.push_control("GPU_CORE_FREQUENCY_MAX_CONTROL",
+                                                     agent_domain,
+                                                     domain_idx), NAN});
+        }
+
+        // We treat energy & time as special cases and only use them at a specific domain.
+        // This is because energy & time are used for for tracking agent behavior/reporting
+        // and do not impact the agent algorithm
+        m_time = {m_platform_io.push_signal("TIME", GEOPM_DOMAIN_BOARD, 0), NAN};
+
+        for (int domain_idx = 0; domain_idx < M_NUM_GPU; ++domain_idx) {
+            m_gpu_energy.push_back({m_platform_io.push_signal("GPU_ENERGY",
+                                    m_platform_io.signal_domain_type("GPU_ENERGY"),
+                                    domain_idx), NAN});
+        }
+
+    }
+
+    // Validate incoming policy and configure default policy requests.
+    void GPUActivityAgent::validate_policy(std::vector<double> &in_policy) const
+    {
+        assert(in_policy.size() == M_NUM_POLICY);
+        double gpu_min_freq = m_platform_io.read_signal("GPU_CORE_FREQUENCY_MIN_AVAIL", GEOPM_DOMAIN_BOARD, 0);
+        double gpu_max_freq = m_platform_io.read_signal("GPU_CORE_FREQUENCY_MAX_AVAIL", GEOPM_DOMAIN_BOARD, 0);
+
+        // Check for NAN to set default values for policy
+        if (std::isnan(in_policy[M_POLICY_GPU_FREQ_MAX])) {
+            in_policy[M_POLICY_GPU_FREQ_MAX] = gpu_max_freq;
+        }
+
+        if (in_policy[M_POLICY_GPU_FREQ_MAX] > gpu_max_freq ||
+            in_policy[M_POLICY_GPU_FREQ_MAX] < gpu_min_freq ) {
+            throw Exception("GPUActivityAgent::" + std::string(__func__) +
+                            "(): GPU_FREQ_MAX out of range: " +
+                            std::to_string(in_policy[M_POLICY_GPU_FREQ_MAX]) +
+                            ".", GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        // Not all gpus provide an 'efficient' frequency signal, and the
+        // value provided by the policy may not be valid.  In this case approximating
+        // f_efficient as midway between F_min and F_max is reasonable.
+        if (std::isnan(in_policy[M_POLICY_GPU_FREQ_EFFICIENT])) {
```

Comment:
Check for LEVELZERO::GPU_CORE_FREQUENCY_EFFICIENT (#2662) and use it if available